### DISCUSSION TITLE
How can I attach a BPF_PROG_TYPE_SK_MSG prog

### DISCUSSION THREAD

**Discussion by fengzx33** [aquasecurity/libbpfgo#334] [Q&A]
I found that libbpfgo haven't an api, which can attach SEC("sk_msg") , what should i do? thanks

example: https://github.com/dafu-wu/istio-tcpip-bypass/blob/796d738babee7da77815ad2819a69d37d8472dea/bpf/bpf_redir.c#L9

**Comment by geyslan** [comment]
Hi @fengzx33, as I mentioned in https://github.com/aquasecurity/libbpfgo/issues/330#issuecomment-1564307452, currently libbpfgo lacks support of `bpf_msg_*` API.

libbpf doesn't use them internally, it just externalizes them.

![image](https://github.com/aquasecurity/libbpfgo/assets/3372117/410de127-906d-4219-8f07-0ef04c692cf7)

> what should i do?

Would you be interested in implementing (wrapping) these uapi in libbpfgo making them available?

**Reply by fengzx33** [reply]
ok，I will give it a try, and if you have any support regarding this matter, please @ me promptly

**Reply by geyslan** [reply]
Awesome. Perhaps these @mozillazg PR can be used as a basis for your effort:

- https://github.com/aquasecurity/libbpfgo/pull/254
- https://github.com/aquasecurity/libbpfgo/pull/205

Any questions, bring them to us.

**Comment by rafaeldtinoco** [comment]
maybe I'm missing context here, but those helpers are to be used by eBPF programs and not by userland (despite being declared in UAPI, because of kernel off-tree compilation needs), no ?

Also, @mozillazg has provided a good example of a socket program in the issue: https://github.com/aquasecurity/libbpfgo/issues/329, which could also serve as basis for an sk_msg example as well.

**Reply by fengzx33** [reply]
@rafaeldtinoco I think you're right. I do need a userland program to attach sk_msg. Can you provide an example for that?
